### ISSUE TITLE
Scipy crash: More than 3*n iterations in LSQ subproblem

### ISSUE THREAD

**Issue by lschaefer** [scikit-hep/pyhf#533]
# Description

Running

```
curl -sL https://gist.githubusercontent.com/lschaefer/e40e397030adc123eefe08bd1502937a/raw/dd758bc53dd9c128f1dc2757412fc20e890a4e3f/gistfile1.txt | pyhf cls
```
causes a scipy crash.

Linking https://github.com/scikit-hep/pyhf/issues/532 because the json files are from the same analysis. Although this issue is reproducible across machines, at least on my laptop and on BNL Tier2.

Any help understanding this would be greatly appreciated!

# Expected Behavior

CLs is computed.

# Actual Behavior

```
ERROR:pyhf.optimize.opt_scipy:     fun: -9.200395594346453
     jac: array([ 2.43040476e+02,  9.36090495e+09,  9.38786444e+09,  9.43490313e+09,
        9.39451435e+09, -5.26518822e-02, -1.72586441e-02, -5.67543507e-01,
       -3.20903540e+00, -2.47903967e+00,  2.42759228e-01,  5.35433960e+01,
        9.33576552e+09,  9.37652329e+09,  9.35860853e+09,  9.29362674e+09,
        1.64408684e-02, -1.86395645e-03, -3.28751087e-01,  1.00917776e+10,
       -1.14739037e+00, -4.55795908e+00, -3.45611572e-02,  9.21915544e+09,
        9.38770048e+09, -9.57393646e-03, -1.00611782e+00,  9.53195990e+09,
       -3.95719528e-01,  1.01782839e+10, -7.23652840e-02, -7.18379021e-02,
       -1.10850811e-01, -1.90390587e-01,  9.72393306e+09,  2.79121399e-01,
        9.74018183e+09,  1.02553057e+10,  1.04391910e+10, -1.80599213e-01,
        9.99519777e+09,  1.12724304e-02, -2.05267906e+00, -7.93562250e+01,
       -3.76801491e-02,  3.47161211e+02,  1.46918297e-02,  3.96013260e-03,
       -1.37634277e-02,  3.77511978e-03, -1.55558586e-02,  3.70304966e+00,
        6.86413288e-01, -8.17442417e-01, -8.03420544e-01, -5.97302914e-01,
       -8.04267406e-01, -2.71524906e-01,  6.07882929e+00, -5.88058472e-01,
       -8.09276581e-01,  9.92220402e-01,  3.20445156e+00, -9.07898903e-01,
        1.32368565e-01, -1.01138067e+00,  4.56362677e+00,  4.81466389e+00,
        4.28961992e+00,  1.02039814e+00, -5.42271137e-01,  1.73331833e+00,
        1.64896917e+00,  2.37052255e+01,  9.11706448e+00, -2.86102295e-06,
        1.90734863e-06, -1.50680542e-04,  7.35980177e+00,  7.62263775e-01,
        9.33338642e-01,  6.17658854e+00, -2.86102295e-06,  2.53453255e-02,
       -2.69205284e+00, -1.13046026e+00,  4.62148619e+00,  4.58211899e-01,
       -1.02890444e+00, -3.38077545e-04,  4.18294859e+00, -1.83269391e+01,
        2.59898133e+01,  1.39190145e+01,  9.69888503e+09, -2.17755508e+00,
        5.36185741e+00, -5.81947803e-01,  3.17038345e+00,  3.00993919e-02,
       -1.75046921e-03,  3.23696136e-01, -1.04735279e+00,  1.18063450e-01,
        2.86102295e-06])
 message: 'More than 3*n iterations in LSQ subproblem'
    nfev: 6647
     nit: 60
    njev: 60
  status: 3
 success: False
       x: array([ 1.01006242e+00, -9.90678874e-02, -1.17963643e-01, -1.37844873e-01,
       -6.98449853e-02, -3.96330466e-03,  7.39755678e-02, -1.93397282e-02,
       -9.87828505e-02, -8.08212324e-02,  5.94643131e-03,  1.08338740e+00,
       -9.75760575e-02, -1.09619133e-01, -1.03931214e-01, -3.47226327e-02,
        7.62332644e-02,  1.19676031e-03, -1.11278292e-02,  1.11318583e+00,
        1.37603640e+00,  9.98961240e-01, -2.10686725e-02, -3.30255504e-02,
       -1.18316933e-01, -8.32864587e-03, -2.90201550e-02,  8.00584944e-02,
       -1.66722008e-02,  3.41431820e-02, -5.25928724e-02, -4.38099421e-02,
       -7.60952050e-02, -5.31209046e-03,  6.04054957e-02,  9.13292888e-03,
        5.90410754e-02,  3.15947458e-02,  2.66769127e-02, -8.86907719e-04,
        4.21897476e-02, -2.37300918e-02,  8.44505090e-01,  9.99638341e-01,
       -2.15329310e-02,  1.00285566e+00, -1.55456272e-02, -6.18387759e-04,
       -7.31369250e-03,  2.93584758e-04, -7.30132311e-03,  1.02217131e+00,
        1.01294659e+00,  9.92073595e-01,  9.78429555e-01,  9.82135574e-01,
        9.79274454e-01,  9.96280017e-01,  1.03834691e+00,  9.87814150e-01,
        9.82649483e-01,  1.00903054e+00,  1.02764223e+00,  9.78817516e-01,
        1.00497239e+00,  9.77477613e-01,  1.02607526e+00,  1.00885676e+00,
        1.00940563e+00,  1.00236888e+00,  9.97745042e-01,  1.01602613e+00,
        1.00581410e+00,  1.05770387e+00,  1.03360020e+00, -1.42361783e-06,
        8.14356635e-07,  9.99891588e-01,  1.02334916e+00,  1.00515582e+00,
        1.00737160e+00,  1.03936287e+00, -1.44026243e-06,  1.00025099e+00,
        9.72354646e-01,  9.95908699e-01,  1.02524878e+00,  1.00448043e+00,
        9.74366699e-01, -1.69020355e-04,  1.01748343e+00,  9.80557844e-01,
        1.01021147e+00,  1.01359197e+00,  1.01660864e+00,  9.90386357e-01,
        1.01425120e+00,  9.94368711e-01,  1.00947605e+00,  1.00016952e+00,
        1.63892462e-03,  1.01213106e+00,  9.75546265e-01,  1.00583143e+00,
        1.36360267e-06])
Traceback (most recent call last):
  File "/Users/lschaefer/miniconda3/bin/pyhf", line 11, in <module>
    sys.exit(pyhf())
  File "/Users/lschaefer/miniconda3/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/lschaefer/miniconda3/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/lschaefer/miniconda3/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/lschaefer/miniconda3/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/lschaefer/miniconda3/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/lschaefer/miniconda3/lib/python3.6/site-packages/pyhf/commandline.py", line 259, in cls
    result = hypotest(testpoi, w.data(p), p, qtilde=is_qtilde, return_expected_set=True)
  File "/Users/lschaefer/miniconda3/lib/python3.6/site-packages/pyhf/utils.py", line 280, in hypotest
    asimov_data = generate_asimov_data(asimov_mu, data, pdf, init_pars, par_bounds)
  File "/Users/lschaefer/miniconda3/lib/python3.6/site-packages/pyhf/utils.py", line 126, in generate_asimov_data
    loglambdav, asimov_mu, data, pdf, init_pars, par_bounds
  File "/Users/lschaefer/miniconda3/lib/python3.6/site-packages/pyhf/optimize/opt_scipy.py", line 43, in constrained_bestfit
    assert result.success
AssertionError
```

@kratsg tried using minuit optimizer instead and got meaningless results instead of a crash:
```
{
    "CLs_exp": [
        1.0,
        1.0,
        1.0,
        1.0,
        1.0
    ],
    "CLs_obs": NaN
}
```

And I made a json from the HistFitter workspace using GenerateJSONOutput.py and got meaningful output:
```

[
    {
        "expectedUpperLimitMinus1Sig": -1.0,
        "upperLimitEstimatedError": -1.0,
        "fitstatus": -1.0,
        "p0d2s": -1.0,
        "p0u2s": -1.0,
        "seed": 0.0,
        "CLsexp": 0.5205284,
        "sigma1": -1.0,
        "failedfit": 0.0,
        "expectedUpperLimitPlus2Sig": -1.0,
        "nofit": 1.0,
        "nexp": -1.0,
        "sigma0": -1.0,
        "clsd2s": 0.1808537,
        "expectedUpperLimit": -1.0,
        "failedstatus": 0.0,
        "xsec": -999007.0,
        "covqual": -1.0,
        "upperLimit": -1.0,
        "p0d1s": -1.0,
        "clsd1s": 0.3166614,
        "failedp0": 0.0,
        "failedcov": 0.0,
        "p0exp": -1.0,
        "p1": 0.2401174,
        "p0u1s": -1.0,
        "excludedXsec": -999007.0,
        "p0": 0.2401174,
        "clsu1s": 0.7602468,
        "clsu2s": 0.9339309,
        "expectedUpperLimitMinus2Sig": -1.0,
        "expectedUpperLimitPlus1Sig": -1.0,
        "mass": 250.0,
        "mode": -1.0,
        "fID": -1.0,
        "dodgycov": 0.0,
        "CLs": 0.5058016
    }
]
```

# Steps to Reproduce
Using python 3.6, pyhf 0.1.2, and scipy 1.3.1.

# Checklist

- [ ] Run `git fetch` to get the most up to date version of `master`
- [ ] Searched through existing Issues to confirm this is not a duplicate issue
- [ ] Filled out the Description, Expected Behavior, Actual Behavior, and Steps to Reproduce sections above or have edited/removed them in a way that fully describes the issue

**Comment by matthewfeickert** [comment]
@kratsg You had mentioned that you think this is related to [this piece of FORTRAN code in SciPy optimize's SLSQP algorithm](https://github.com/scipy/scipy/blob/0e7bd3b789c4e4c8ce5517e4982ef6d354771f97/scipy/optimize/slsqp/slsqp_optmz.f#L1668). Once you've had more time to look at it can you also post a short summary of your thoughts here?

**Comment by lschaefer** [comment]
This may be an issue with my HF config and not a bug in pyhf or its dependences. The pulls of the background only fit looks ok: 
[fit_parameters_bkg.pdf](https://github.com/diana-hep/pyhf/files/3557245/fit_parameters_bkg.pdf)

But the pulls of the exclusion fit look pretty bad, with a lot of over- and under-constraints:
[fit_parameters_excl.pdf](https://github.com/diana-hep/pyhf/files/3557247/fit_parameters_excl.pdf)

So maybe the systematics aren't being treated correctly in my exclusion config. I'll sort this out and update here soon.
-Leigh

**Comment by matthewfeickert** [comment]
@lschaefer I'm going to close this as with a modern release of `pyhf` (`v0.4.X`) this passes

```
$ curl -sL https://gist.githubusercontent.com/lschaefer/e40e397030adc123eefe08bd1502937a/raw/dd758bc53dd9c128f1dc2757412fc20e890a4e3f/gistfile1.txt | pyhf cls
/home/feickert/Code/GitHub/pyhf/src/pyhf/tensor/numpy_backend.py:263: RuntimeWarning: invalid value encountered in log
  return n * np.log(lam) - lam - gammaln(n + 1.0)
{
    "CLs_exp": [
        0.43493602299316964,
        0.578032306619249,
        0.7411647536427226,
        0.8896137693846295,
        0.9746836971668921
    ],
    "CLs_obs": 0.5685073288269009
}
```

I doubt this Issue is still of interest to you, given how old it is (sorry!), but I'm just going through the backlogs. I also am not 100% sure where this got patched (I can check if anyone is really curious) but my guess is that this got fixed either alongside Issue #533 or in release [`v0.4.2`](https://github.com/scikit-hep/pyhf/releases/tag/v0.4.2) (as the result of PR #796).

**Comment by kratsg** [comment]
> I doubt this Issue is still of interest to you, given how old it is (sorry!), but I'm just going through the backlogs. I also am not 100% sure where this got patched (I can check if anyone is really curious) but my guess is that this got fixed either alongside Issue #532 or in release [`v0.4.2`](https://github.com/scikit-hep/pyhf/releases/tag/v0.4.2) (as the result of PR #796).

Identifying this now for easier reference later. First run:

```
(pyhf-dev) Lord Stark:~/pyhf (master|BISECTING)$ git bisect fixed master
(pyhf-dev) Lord Stark:~/pyhf (master|BISECTING)$ git bisect unfixed v0.1.2
Bisecting: 106 revisions left to test after this (roughly 7 steps)
[ebf7ece42a4018d8fdcd71767b2410d4a22ba64a] feat: Migrate to TensorFlow v2.0 (#541) 
```

then I'll keep running `pip install -e .; pyhf cls test.json` to identify if fixed `git bisect fixed` or unfixed `git bisect unfixed`. (Note: Need to constantly re-install the CLI as it was changed quite a lot over this large of a range of current master ~v0.4.4 since v0.1.2). This gets me

```
$ git bisect fixed
9b93f9d243ad71e9bde7f1428242df661be5c97d is the first fixed commit
commit 9b93f9d243ad71e9bde7f1428242df661be5c97d
Author: Lukas <lukas.heinrich@gmail.com>
Date:   Fri Sep 6 11:35:21 2019 +0200

    Enable full batch evaluation (#546)
    
    * Add ability to do batched evaluation
    * pdf.logpdf(pars, data), where pars.shape = (B,N) with B==pdf.batch_size and N==len(pdf.config.suggested_config())
    * Add test to expose bug in constrained_by_poisson.expected_data
    * Add fix for bug in constrained_by_poisson.expected_data
    * Relax tolerance in validation (suspect it's due to order of moving python arrays to tensors)
    * Add arg in hepdata_like to return batched model
    * Rename 'thenom' to 'nominal_rates'

:040000 040000 db543851fe6da7c17922c3012ab64a4963ce0a41 1b1936fc626fabb834d0becafd1168def7200195 M	docs
:040000 040000 eba5550cb2fc55e85f7b734047c245b3691b0518 7cdb1441e7f31cffdd9e5d7c26ad7cdbd8a09a2d M	src
:040000 040000 c993a70ca7dad2a48ffda2f0d1b860490c29ff34 0b1583772cddad1b04faf8fee0bd9e53e7750a02 M	tests
```

Likely, it's due to the fix here (https://github.com/scikit-hep/pyhf/pull/546/files#diff-0252cafa7a020d64286afec50127db84R41-R42)
